sycl/doc/extensions/experimental/SYCL_EXT_ONEAPI_GRAPH.asciidoc

```diff
@@ -0,0 +1,274 @@
+GitHub does not render image icons.
+
+This extension is written against the SYCL 2020 revision 4 specification. All
+references below to the "core SYCL specification" or to section numbers in the
+SYCL specification refer to that revision.
+
+NOTE: This extension is experimental: interfaces are subject to change later.
+
+== Introduction
+
+This extension introduces an interface that enables a lazy execution and easy replay of a kernel graph by separating
+Its definition and execution. 
+
+== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification section 6.3.3 "Feature test macros".  Therefore, an
+implementation supporting this extension must predefine the macro
+`SYCL_EXT_ONEAPI_GRAPH` to one of the values defined in the table below.
+Applications can test for the existence of this macro to determine if the
+implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's APIs the implementation supports.
+
+Table 1. Values of the `SYCL_EXT_ONEAPI_GRAPH` macro.
+[%header,cols="1,5"]
+|===
+|Value |Description
+|1     |Initial extension version. Base features are supported.
+|===
+
+== SYCL Graph Terminology
+
+Table 2. Terminology.
+|===
+|Concept|Description
+|graph| Class that stores structured work units and their dependencies
+|node| The unit of work. Can have different attributes.
+|edge| Dependency between work units. Happens before relation.
+|===
+
+== Node
+
+Node is a class that encapsulates tasks like SYCL kernel functions or host tasks for deferred execution.
+A graph has to be created first, the structure of a graph is defined second by adding nodes and edges.
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+  class node{
+  };
+}
+----
+
+NOTE:
+
+== Edge
+
+A dependency between two nodes representing a happens before relationship. `sender` and `receiver` may be associated to different graphs.
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+	// Adding dependency between two nodes.
+    	void make_edge(node sender, node receiver);
+}
+----
+
+== Graph
+
+Graph is a class that represents a directed acyclic graph of nodes. 
+A graph can have different states, can be nested, can have multiple root nodes that are scheduled for execution first and multiple leaf nodes that are scheduled for execution last. The execution of a graph has been completed when all leaf node tasks have been completed.
+Member functions as listed in Table 2 and 3 can be used to add nodes to a graph.
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+  enum class graph_state{
+    modifiable,
+    executable
+  };
+
+  template<graph_state State>
+  class graph {
+  public:
+    operator graph<graph_state::executable>();
+  };
+  
+  graph<graph_state::modifiable> make_graph();
+  
+  graph<graph_state::modifiable> compile(const graph<graph_state::executable> Graph);
```

Comment:
why the status of input graph is executable, and the return value is modifiable?

In my understanding, function compile will compile the graph into hardware instructions with device memory allocated, etc. And so, the result of compile is executable.

Reply:
as for the compile, build APIs, we may reference to the Khronos OpenGL/OpenCL APIs. And NNAPI (android neural network API) is also a good reference for the graph, see https://developer.android.com/ndk/guides/neuralnetworks.

Reply:
> why the status of input graph is executable, and the return value is modifiable?
> 
> In my understanding, function compile will compile the graph into hardware instructions, etc. And so, the result of compile is executable.

Good catch, attributes have to be swapped for `compile`. 👍